ad7768-1fmcz/ad77681.c

```diff
@@ -486,21 +1105,555 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
+ * @param status		Structure with all satuts bits
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad77681_status(struct ad77681_dev *dev,
+		       struct ad77681_status_registers *status)
+{
+	int32_t ret;
+	uint8_t buf[3];
+
+	/* Master status register */
+	ret = ad77681_spi_reg_read(dev, AD77681_REG_MASTER_STATUS,buf);
+	status->master_error = buf[1] & AD77681_MASTER_ERROR_MSK;
+	status->adc_error = buf[1] & AD77681_MASTER_ADC_ERROR_MSK;
+	status->dig_error = buf[1] & AD77681_MASTER_DIG_ERROR_MSK;
+	status->adc_err_ext_clk_qual = buf[1] & AD77681_MASTER_DIG_ERR_EXT_CLK_MSK;
+	status->adc_filt_saturated = buf[1] & AD77681_MASTER_FILT_SAT_MSK;
+	status->adc_filt_not_settled = buf[1] & AD77681_MASTER_FILT_NOT_SET_MSK;
+	status->spi_error = buf[1] & AD77681_MASTER_SPI_ERROR_MSK;
+	status->por_flag = buf[1] & AD77681_MASTER_POR_FLAG_MSK;
+	/* SPI diag status register */
+	ret |= ad77681_spi_reg_read(dev, AD77681_REG_SPI_DIAG_STATUS, buf);
+	status->spi_ignore = buf[1] & AD77681_SPI_IGNORE_ERROR_MSK;
+	status->spi_clock_count = buf[1] & AD77681_SPI_CLK_CNT_ERROR_MSK;
+	status->spi_read_error = buf[1] & AD77681_SPI_READ_ERROR_MSK;
+	status->spi_write_error = buf[1] & AD77681_SPI_WRITE_ERROR_MSK;
+	status->spi_crc_error = buf[1] & AD77681_SPI_CRC_ERROR_MSK;
+	/* ADC diag status register */
+	ret |= ad77681_spi_reg_read(dev, AD77681_REG_ADC_DIAG_STATUS,buf);
+	status->dldo_psm_error = buf[1] & AD77681_ADC_DLDO_PSM_ERROR_MSK;
+	status->aldo_psm_error = buf[1] & AD77681_ADC_ALDO_PSM_ERROR_MSK;
+	status->ref_det_error = buf[1] & AD77681_ADC_REF_DET_ERROR_MSK;
+	status->filt_sat_error = buf[1] & AD77681_ADC_FILT_SAT_MSK;
+	status->filt_not_set_error = buf[1] & AD77681_ADC_FILT_NOT_SET_MSK;
+	status->ext_clk_qual_error = buf[1] & AD77681_ADC_DIG_ERR_EXT_CLK_MSK;
+	/* DIG diag status register */
+	ret |= ad77681_spi_reg_read(dev, AD77681_REG_DIG_DIAG_STATUS,buf);
+	status->memoy_map_crc_error = buf[1] & AD77681_DIG_MEMMAP_CRC_ERROR_MSK;
+	status->ram_crc_error = buf[1] & AD77681_DIG_RAM_CRC_ERROR_MSK;
+	status->fuse_crc_error = buf[1] & AD77681_DIG_FUS_CRC_ERROR_MSK;
+
+	return ret;
+}
+
+/**
+ * Initialize the device and ADC status structure.
  * @param device - The device structure.
  * @param init_param - The structure that contains the device initial
  * 					   parameters.
+ * @param status	 - First status read from ADC
  * @return 0 in case of success, negative error code otherwise.
  */
 int32_t ad77681_setup(struct ad77681_dev **device,
-		      struct ad77681_init_param init_param)
+		      struct ad77681_init_param init_param,
+		      struct ad77681_status_registers **status)
 {
 	struct ad77681_dev *dev;
-	int32_t ret = 0;
+	struct ad77681_status_registers *stat;
+	int32_t ret;
+	uint8_t scratchpad_check = 0xAD;
 
 	dev = (struct ad77681_dev *)malloc(sizeof(*dev));
 	if (!dev) {
-		return -1;
+		return FAILURE;
+	}
+
+	stat = (struct ad77681_status_registers *)malloc(sizeof(*stat));
+	if (!stat) {
+		return FAILURE;
```

Comment:
`free(dev);` before returning FAILURE

Reply:
Memory leak build error was fixed when returning -1 instead of failure.